clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/FleetController.java

```diff
@@ -52,6 +55,9 @@
 
     private static final Logger log = Logger.getLogger(FleetController.class.getName());
 
+    // Components of different DI graph generation shan't run() simultaneously.
+    private static final Map<String, Lock> controllerLocks = new ConcurrentHashMap<>();
```

Comment:
Note that this will only work as long as the FleetController is part of an installed bundle. If it were in a different bundle, as all user provided code is this approach will not work as you will get one map per classloader/bundle.

I do seem to recall that we have another mechanism for this that will also work across bundles. I think that @gv @bjorncs @hmusum or @bratseth might know the details.

Reply:
Yes, it only works for installed bundles, and within the same JVM—we potentially have this problem when new CC-0 starts up before old CC-0 retires, as well, but that probably shouldn't happen in practice. 

We have `ClassLocking`, which uses class names as keys, and this will work across bundles. It can't be used directly here, however, as the key needs to include cluster name and ID.